packages/piro/test/Main_AnalysisDriver_Tpetra.cpp

```diff
@@ -0,0 +1,176 @@
+  bool doAll = (argc==1);
+  if (argc>1) doAll = !strcmp(argv[1],"-v");
+
+  Piro::SolverFactory solverFactory;
+
+  for (int iTest=0; iTest<2; iTest++) {
+
+    if (doAll) {
+      switch (iTest) {
+       case 0: inputFile="input_Analysis_ROL_Tpetra.xml"; break;
+       case 1: inputFile="input_Analysis_ROL_AdjointSensitivities_Tpetra.xml"; break;
+       case 2: inputFile="input_Analysis_OptiPack.xml"; break;
+       case 3: inputFile="input_Analysis_MOOCHO.xml"; break;
+       default : std::cout << "iTest logic error " << std::endl; exit(-1);
+      }
+    }
+    else {
+      inputFile=argv[1];
+      iTest = 999;
+    }
+
+    if (Proc==0)
+     std::cout << "===================================================\n"
+          << "======  Running input file "<< iTest <<": "<< inputFile <<"\n"
+          << "===================================================\n"
+          << std::endl;
+
+    try {
+
+      // Create (1) a Model Evaluator and (2) a ParameterList
+      const RCP<Thyra::ModelEvaluator<double>> Model = rcp(new MockModelEval_A_Tpetra(appComm));
+
+      // BEGIN Builder
+      const RCP<Teuchos::ParameterList> appParams = rcp(new Teuchos::ParameterList("Application Parameters"));
+      Teuchos::updateParametersFromXmlFile(inputFile, Teuchos::ptr(appParams.get()));
+
+      const RCP<Teuchos::ParameterList>  piroParams = Teuchos::sublist(appParams,"Piro");
+      Teuchos::ParameterList& analysisParams = appParams->sublist("Analysis");
+
+
+      Stratimikos::DefaultLinearSolverBuilder linearSolverBuilder;
+      typedef Thyra::PreconditionerFactoryBase<double>              Base;
+      typedef Thyra::Ifpack2PreconditionerFactory<Tpetra_CrsMatrix> Impl;
+      linearSolverBuilder.setPreconditioningStrategyFactory(
+          Teuchos::abstractFactoryStd<Base, Impl>(), "Ifpack2");
+      Stratimikos::enableMueLu<int, int>(linearSolverBuilder);
+
+      const Teuchos::RCP<Teuchos::ParameterList> stratList = Piro::extractStratimikosParams(piroParams);
+      linearSolverBuilder.setParameterList(stratList);
+
+
+      const RCP<Thyra::LinearOpWithSolveFactoryBase<double>> lowsFactory =
+          createLinearSolveStrategy(linearSolverBuilder);
+
+      RCP<Thyra::ModelEvaluator<double>> ModelWithSolve = rcp(new Thyra::DefaultModelEvaluatorWithSolveFactory<double>(
+          Model, lowsFactory));
+
+      // Use these two objects to construct a Piro solved application
+      // Thyra::ModelEvaluator is the base class of all Piro::Epetra solvers
+      const RCP<Thyra::ModelEvaluatorDefaultBase<double>> piro = solverFactory.createSolver(piroParams, ModelWithSolve);
+      // END Builder
+
+      // Call the analysis routine
+      RCP<Thyra::VectorBase<double>> p;
+      status = Piro::PerformAnalysis(*piro, analysisParams, p);
+
+      if (Teuchos::nonnull(p)) {
+        // Can post-process results here
+        if (Proc==0) {
+          std::cout << "\nPiro_AnalysisDrvier:  Optimum printed above has exact soln = {1,3}" << std::endl;
```

Comment:
Are the values of the optimum checked somewhere in this test? Why just printed to screen here? Could we change this driver to fail the test if the values aren't correct?

Reply:
This makes sense. I ported this driver from the Epetra version one, but it makes sense to have an error check. I'll put the error check in the Epetra driver as well.